### PR TITLE
Change string representation to support PIC

### PR DESCRIPTION
**Pull request by jberdine**
Change string representation to use bit 1 of the hash instead of the high bit to distinguish strings.

### CODE REVIEW

**Comment by jberdine** [comment]
(CI is expected to fail, these changes will require a bootstrap and docker image dance.)

Just putting this up so others can see some of what is cooking. Individual commits should be readable more easily than the whole stack at once.

**Comment by jberdine** [comment]
These changes are not all relevant to the string representation, but also include other fixes and cleanup that I encountered while working on strings.

**Comment by mbouaziz** [comment]
Is there a prefix that can be reviewed and merged?

**Comment by jberdine** [comment]
Yes. A number of these changes can't be tested in CI due to needing a more manual bootstrap procedure, and some will break CI for other PRs until the base docker image is updated. So I have been accumulating this to amortize the churn and risk. But since I'm not the only one working on these files now it makes sense to expidite merging more.

**Comment by jberdine** [comment]
I need to go through and check that these commits don't undo @mbouaziz's recent improvements, and redo bootstraps after rebasing. But otherwise these commits up to but not including `[compiler] Change string hash tag bit from 31 to 3 in generated code` are ready for review and can be merged after that and another round of rebasing and testing.

The `[compiler] Change string hash tag bit from 31 to 3 in generated code` and following commits show the current state that works in native but not wasm. Using bit 3 for the string hash tag works in native since vtable pointers are 8-aligned, but not in wasm as there they are only 4-aligned. Using bit 2 does not currently work. Some other code is using bit 2 it seems. Bit 1 is used by copy and intern.

File: compiler/src/AsmOutput.sk
Comment (mbouaziz) [inline compiler/src/AsmOutput.sk:3428]: `StringConcat` is produced at two places:
- in `intrinsics.sk`, `.IntrinsicsInliner.maybeInlineIntrinsic` for `String::concat` which is a method that takes only one parameter and hence will produce only `SKIP_String_concat2`
- in `rewrite.sk`, `.Rewrite.emitStringConcat`, which is never called.

Hence only the binary version is ever produced.

The suffix for the varargs version should be `"N"`, not `""`.

File: prelude/src/stdlib/other/System.sk
Comment (mbouaziz) [inline prelude/src/stdlib/other/System.sk:119]: Why doesn't the compiler give warnings for duplicates?

File: compiler/runtime/runtime64_specific.cpp
Comment (mbouaziz) [inline compiler/runtime/runtime64_specific.cpp]: This should certainly be a static constant

File: compiler/runtime/runtime64_specific.cpp
Comment (mbouaziz) [inline compiler/runtime/runtime64_specific.cpp]: Sorry, I touched this function in #148

File: prelude/src/native/StringIterator.sk
Comment (mbouaziz) [inline prelude/src/native/StringIterator.sk:16]: Is there any native `StringIterator` implementation?

File: compiler/runtime/runtime.c
Comment (mbouaziz) [inline compiler/runtime/runtime.c]: Should it be written
`(sk_string_t*)(obj - offsetof(sk_string_t, data))`?

**Review by mbouaziz — COMMENTED** [review]
LGTM though I'm not sure I fully grasp the differences between `@cpp_extern`, `@cpp_runtime`, `@intrinsic` + what is in the preamble.

I think it's worth merging the prefix that works

**Comment by jberdine** [comment]
Re the differences between `@cpp_extern`, `@cpp_runtime`, `@intrinsic` + what is in the preamble:

- `intrinsic` is for functions that the compiler itself knows about and will lower to some other code rather than just emitting a call to the named function. Sometimes a function is marked `intrinsic` if it is transformed to something by the compiler only conditionally or for some targets, in which case a default implementation is also needed. There are currently cases (e.g. `clzImpl`) where such default implementations exist but are never used. It's a TODO to clean them up.
- `cpp_runtime` and `cpp_extern` are very similar in that they declare a function for skip code and delegate the implementation to the runtime system. If a name is given, that is the C linkage name that will be used for the implementation, if no name is given, the default encoding of skip names to C linkage names is used (e.g. `SKIP_<module>_<class>_<function>`). The difference between `cpp_runtime` and `cpp_extern` is that compiler will emit a declaration for `cpp_extern` functions but will not for `cpp_runtime` ones. Therefore, the `cpp_runtime` ones need to be declared manually in the preamble. This is useful to e.g. give different but compatible types in the llvm declaration than would be generated.

File: compiler/src/AsmOutput.sk
Comment (jberdine) [reply compiler/src/AsmOutput.sk:3428]: Yes, only the binary version is produced. I left it for later to decide whether to reintroduce the optimization and vararg version or to remove the support from the compiler. It was removed in https://github.com/SkipLabs/skfs/pull/142 but this seems to me to be a tangential change that we might still want. So I left it for now.

Also, my understanding is that the vararg version is a currently non-existent runtime function named `SKIP_String_concat` while the existing `SKIP_String_concatN` is a different function that concats an array of strings. I don't think that and the vararg version are / can be the same.

File: prelude/src/stdlib/other/System.sk
Comment (jberdine) [reply prelude/src/stdlib/other/System.sk:119]: *shrug*

File: compiler/runtime/runtime64_specific.cpp
Comment (jberdine) [reply compiler/runtime/runtime64_specific.cpp]: That seems like a good idea.

File: prelude/src/native/StringIterator.sk
Comment (jberdine) [reply prelude/src/native/StringIterator.sk:16]: There is `SKIP_String_StringIterator__substring` in string.c, but I think that is all.

File: compiler/runtime/runtime.c
Comment (jberdine) [reply compiler/runtime/runtime.c]: I thought about that but held off for now since it would take a bit more work to be able to do it consistently that way for all of strings, arrays, and general skip objects. I was also thinking that the compiler code produces values with exact knowledge about the layout, and writing the runtime code using `offsetof` gives the impression that things are robust with respect to the compiler making different decisions about struct layout such as adding padding, when in fact that is not true. But the runtime code is much easier to read using actual C types and just have in mind that we are still sensitive to the exact layout the compiler currently uses. So after your other changes, it would make sense to change this to use `offsetof` as well.

**Comment by jberdine** [comment]
To try to repro the failure I see on my laptop but not on an aws box:

1. pull down this branch (https://github.com/jberdine/skdb/tree/string)
2. checkout the `[compiler] Change string hash tag bit from 31 to 31 and 1 in generated code` commit: `git checkout string~3`
3. build _and install_ the stage1 compiler, I run: `set -a; SKARGO_PROFILE=release; STAGE=1; PATH=$PWD/compiler/stage1/bin:$PWD/compiler/stage0/bin:$PATH; set +a; make clean && make -C compiler clean install`
4. checkout the tip of this branch: `git checkout string`
5. build the stageC compiler, I run: `set -a; SKARGO_PROFILE="dev -v"; STAGE=C; PATH=$PWD/compiler/stageC/bin:$PWD/compiler/stageB/bin:$PWD/compiler/stageA/bin:$PATH; set +a; make clean && make -C compiler clean default`

Note that in the last step, I add `-v` to `SKARGO_PROFILE` in order to put `skargo` in verbose mode, which lets the error output get to the terminal.

The bad behavior I see is that when executing `skc --data stageC/bin/state.db --export-function-as main=skip_main -o stageC/bin/skc ...`, after the `clang` subprocess terminates, something like the following error message is printed:
```
bit 1 but not 31 set: 0x453f4c9c
make[1]: *** [Makefile:95: stageC/bin/skc] Error 67
```
This comes from line 43 or runtime.c.

**Comment by jberdine** [comment]
This is ready now if anyone wants to make a review pass. I think/hope that the individual commits are each independeltly readable. The CI is expected to fail, the base image will need to be updated.

To test this I ran the following to test each commit in the stack one-by-one:
```
GIT_SEQUENCE_EDITOR=: git rebase -ix "dx 'set -a; SKARGO_PROFILE=dev; OLEVEL=-O0; STAGE=D; PATH=stageD/bin:\$PATH; set +a; make clean && make -C compiler clean install && make test-skfs'" main
```
where `dx` is my script to invoke a command in a docker container.
(Actually this was done only starting from the commit that adds the `stageD` build).

This process checks that a working compiler is produced at each step, when using the parent commit's compiler installed onto the host. This is true for the stack except for the step between `[runtime] Change to use bit 1 to distinguish strings` and `[compiler] Change to use bit 1 to distinguish strings in generated code` where the installed compiler generates code expecting the new runtime system, which only comes with the next commit. For this step, skc was manually installed onto the host with `set -a; SKARGO_PROFILE=release; STAGE=D; PATH=stageD/bin:$PATH; set +a; cd compiler; install -m 755 stage$STAGE/bin/skc /usr/local/bin`.

FInally, the tip was tested with a full `make test` using the tip compiler and runtime.

**Comment by jberdine** [comment]
Note that the bootstrap commits, except for the last one, need not be committed. They mark points where going further in the stack requires the preceding runtime changes. I will filter them out before merging.

File: compiler/Makefile
Comment (mbouaziz) [inline compiler/Makefile:82]: If you forget to clean the stageD binaries, this would use the latest stageD skargo to build the new one.

If you want to use skargo from the `PATH`, you should resolve the path to skargo before adding `stageD/bin/` to the `PATH`

Comment (jberdine) [reply compiler/Makefile:82]: Yes, this will use a left-over stageD/bin/skargo if present. My experience building the compiler etc. in general has led me to always clean before a build. That's why my test procedure has e.g. `make clean && make -C compiler clean install && make test-skfs`. The current state of the Makefiles is, I think, that the declared dependencies suffice to determine an order in which to run the build commands from scratch, but there are many missing dependencies that would be needed to detect changes robustly. So far I have not spent the time to make non-clean builds robust. That is another project, and realistically I don't think that there is much to save since the build time is highly dominated by running skc and clang on pretty much all the files if anything has changed.

**Review by mbouaziz — APPROVED** [review]
LGTM :+1:

**Comment by mbouaziz** [comment]
CI is red though

**Comment by jberdine** [comment]
> CI is red though

Yes, CI can't handle PRs like this. It tries to build the current skdb with the old compiler and the current prelude.

I will strip out the extra bootstrap commits, land this, then rebuild the image for CI.

**Pull request merged**
